tests: added fixture with energy usage instead of example

kkucharc opened the pull request:
All tests which requires data from energy usages will be loaded and cleaned up after each test thanks to fixture.

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TEST PLAN
<!--- What steps should be taken to verify the changes -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

kkucharc committed:
Added energy usage fixture. Updated dashboard utils and unicode fixture with new method parameters.

kkucharc committed:
Add energy fixture to tests/access_tests.py

kkucharc committed:
Add energy fixture to tests/core_tests.py

kkucharc committed:
Add energy fixture to tests/dashboard_tests.py

kkucharc committed:
Add energy fixture to tests/datasets/api_tests.py

kkucharc committed:
Add energy fixture to tests/db_engine_specs/base_engine_spec_tests.py

kkucharc committed:
Add energy fixture to tests/import_export_tests.py

kkucharc committed:
Add energy fixture to tests/model_tests.py

kkucharc committed:
Add energy fixture to tests/query_context_tests.py

kkucharc committed:
Add energy fixture to tests/security_tests.py

kkucharc committed:
Add energy fixture to tests/charts/api_tests.py

kkucharc committed:
Changed formatting of slices' parameters in energy usage fixture

kkucharc committed:
Removed loading energy udage data from test conf file

kkucharc committed:
Add energy fixture to tests/databases/api_tests.py

willbarrett commented on tests/fixtures/energy_dashboard.py:
```diff
@@ -0,0 +1,357 @@
+    table_name = "energy_usage"
+    df = _get_dataframe()
+    with app.app_context():
+        yield _create_energy_table(df, table_name)
```
should this be:
```python
_create_energy_table(df, table_name)
yield
_cleanup()
```
?

willbarrett reviewed (approved):
Mostly looks good, one question.

villebro commented on tests/fixtures/energy_dashboard.py:
```diff
@@ -0,0 +1,357 @@
+        {
+            "source": "Rail - Ship and Other Transport",
+            "target": "Carbon Dioxide",
+            "value": "2.5",
+        },
+        {
+            "source": "Residential Buildings",
+            "target": "Carbon Dioxide",
+            "value": "10.2",
+        },
+        {"source": "Rice Cultivation", "target": "Methane", "value": "1.5"},
+        {"source": "Road", "target": "Carbon Dioxide", "value": "10.5"},
+        {"source": "T and D Losses", "target": "Carbon Dioxide", "value": "2.2"},
+        {"source": "Transportation", "target": "Air", "value": "1.7"},
+        {
+            "source": "Transportation",
+            "target": "Rail - Ship and Other Transport",
+            "value": "2.5",
+        },
+        {"source": "Transportation", "target": "Road", "value": "10.5"},
+        {
+            "source": "Unallocated Fuel Combustion",
+            "target": "Carbon Dioxide",
+            "value": "3.0",
+        },
+        {"source": "Unallocated Fuel Combustion", "target": "Methane", "value": "0.4"},
+        {
+            "source": "Unallocated Fuel Combustion",
+            "target": "Nitrous Oxide",
+            "value": "0.4",
+        },
+        {"source": "Waste", "target": "Landfills", "value": "1.7"},
+        {"source": "Waste", "target": "Waste water - Other Waste", "value": "1.5"},
+        {"source": "Waste water - Other Waste", "target": "Methane", "value": "1.2"},
+        {
+            "source": "Waste water - Other Waste",
+            "target": "Nitrous Oxide",
+            "value": "0.3",
+        },
+    ]
```
I wonder if we should utilize the existing data fetching logic from the examples data? See `superset/examples/helpers.py` and `superset/examples/energy.py:load_energy()`. I think we could potentially cache the data on disk, so it only needs to be fetched once if it's already there (I don't expect this data to change that often).

kkucharc replied:
@villebro to be honest, I imagine as developer who wants write/run tests (especially unit tests), I would like have data that isn't fetched from web - so I don't need internet to just run tests. Actually this is why I didn't use this fetching on purpose. Also if want use fixtures, probably we would like to cleanup after each test. I think it would extend test setup significantly then. But what do you think about just moving it to separate file (as energy_usage.json) or just shortening collection (if it's possible)?

Actually I had an idea to generate this data with some random values of energy sources, targets and values. Apparently it requires some structure (as far as I remember it complained about some cycles). Maybe you know how this data should be structured so I could generate it with loop?

villebro replied:
@kkucharc regarding having fixtures that are generated as opposed to fetched, I do agree that it makes sense. Some time ago I added some `DataFrame` fixtures (before `pytest`, hence they're not done the elegant way):

https://github.com/apache/incubator-superset/blob/ac498ddd2310f1692d6bf204042a795bd5bcf897/tests/fixtures/dataframes.py#L106-L116

This and others in this file are used to test some `DataFrame` transformation functions in `utils/pandas_postprocessing.py`, where I can aggregate/sort/filter the `DataFrame` and then assert that the result is as expected. What makes this convenient is that the dataset is very simple, yet provides enough complexity to be able to be able to ensure that the functions work as expected.

So, for this case, perhaps we should just completely move off the energy datasets, and start using some programmatically generated dataset, and then start applying that template to all other examples datasets that are used throughout the backend tests.

@dpgaspar @betodealmeida @hughhhh do you have thoughts about how to proceed here?

willbarrett replied:
My 2 cents would be to migrate away from the energy test fixture entirely and move to random data. I agree with Kasha's statement that relying on a data load from the network isn't a great thing for a test suite. I'd prefer for us to move to testing with more-generic and randomly generated data.

villebro reviewed (commented):
Great work again @kkucharc! 🌮 Some comments, curious to hear your thoughts!

dpgaspar reviewed (commented):
Looks really good, excited to see this. Would be great if we could leverage the existing data download ( @villebro suggestion )

hughhhh commented on tests/core_tests.py:
```diff
@@ -59,6 +61,7 @@
 from superset.utils import core as utils
 from superset.views import core as views
 from superset.views.database.views import DatabaseView
+from tests.utils import read_fixture
```
`read_fixture` isn't being used

kkucharc commented on tests/charts/api_tests.py:
```diff
@@ -691,7 +694,7 @@ def test_get_charts_custom_filter(self):
             self.assertEqual(item["slice_name"], expected_response[index]["slice_name"])
             self.assertEqual(item["viz_type"], expected_response[index]["viz_type"])
 
-    @pytest.mark.usefixtures("load_charts")
+    @pytest.mark.usefixtures("load_energy_table_with_slice", "load_charts")
```
Sure, sounds good! 👍 👌

kkucharc commented on tests/db_engine_specs/base_engine_spec_tests.py:
```diff
@@ -14,19 +14,22 @@
+from ..fixtures.energy_dashboard import load_energy_table_with_slice
 from ..fixtures.pyodbcRow import Row
 
+from tests.test_app import app  # isort:skip
+
```
You are right, that was the case, thanks!

kkucharc committed:
Fixes after review: removed isort:skip, load_charts->load_energy_charts, removed unused import.

kkucharc committed:
Added energy fixture to tests/charts/commands_tests.py and retrieving proper Slice by name

kkucharc committed:
Fixed charts/api_tests.py to use energy_usage from fixtures

kkucharc committed:
Fixed datasets/commands_tests.py to retrieve dataset by name and use energy_usage fixture

kkucharc committed:
Changed energy usage data to generated data and fixed chart tests which was checking energy usage data

kkucharc commented:
Sorry for the delay in the topic (I needed switch priorities).

@villebro I added generating data as you suggested in the [comment](https://github.com/apache/incubator-superset/pull/11308#discussion_r511915875). 
I also fixed all tests that were new and not passing with this fixture. I will rebase the conflict after review.
cc: @dpgaspar @willbarrett

willbarrett reviewed (approved):
LGTM after rebase!

villebro reviewed (approved):
LGTM after fixing conflict
